pkg/controller/pubsub/pubsub_init.go

```diff
@@ -0,0 +1,134 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+
+	constraintclient "github.com/open-policy-agent/frameworks/constraint/pkg/client"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/externaldata"
+	"github.com/open-policy-agent/gatekeeper/pkg/expansion"
+	"github.com/open-policy-agent/gatekeeper/pkg/logging"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation"
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub"
+	psClient "github.com/open-policy-agent/gatekeeper/pkg/pubsub/client"
+	"github.com/open-policy-agent/gatekeeper/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/pkg/watch"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller").WithValues(logging.Process, "pubsub_controller")
+
+type Adder struct {
+	PubsubSystem *pubsub.System
+}
+
+func (a *Adder) Add(mgr manager.Manager) error {
+	if len(psClient.Providers()) == 0 {
+		return nil
+	}
+	r := newReconciler(mgr, a.PubsubSystem)
+	return add(mgr, r)
+}
+
+func (a *Adder) InjectOpa(_ *constraintclient.Client) {}
+
+func (a *Adder) InjectWatchManager(_ *watch.Manager) {}
+
+func (a *Adder) InjectControllerSwitch(_ *watch.ControllerSwitch) {}
+
+func (a *Adder) InjectTracker(_ *readiness.Tracker) {}
+
+func (a *Adder) InjectMutationSystem(_ *mutation.System) {}
+
+func (a *Adder) InjectExpansionSystem(_ *expansion.System) {}
+
+func (a *Adder) InjectProviderCache(_ *externaldata.ProviderCache) {}
+
+func (a *Adder) InjectPubsubSystem(pubsubSystem *pubsub.System) {
+	a.PubsubSystem = pubsubSystem
+}
+
+type Reconciler struct {
+	client.Client
+	scheme *runtime.Scheme
+	system *pubsub.System
+}
+
+func newReconciler(mgr manager.Manager, system *pubsub.System) *Reconciler {
+	return &Reconciler{
+		Client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+		system: system,
+	}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	c, err := controller.New("pusbub-config-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(
+		&source.Kind{Type: &corev1.Secret{}},
```

Comment:
Where is "Secret" currently used in this controller?

IMO the need to watch secondary resources, like secrets, is a good reason to have extension authors write their own controllers, unless we're comfortable being prescriptive about what kinds of credential types are acceptable and how they should be packaged/referenced.